plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAMonitor.java

```diff
@@ -35,121 +35,221 @@
+                            removedPools.add(uuid);
+                            continue;
                         }
 
-                        removedPools.add(uuid);
-                        continue;
                     }
 
-                    s_logger.debug(String.format("Found NFS storage pool [%s] in libvirt, continuing.", uuid));
+                    String result = null;
+                    for (int i = 1; i <= _heartBeatUpdateMaxTries; i++) {
+                        Script cmd = createHeartBeatCommand(primaryStoragePool, hostPrivateIp, true);
+                        result = cmd.execute();
+
+                        s_logger.debug(String.format("The command (%s), to the pool [%s], has the result [%s].", cmd.toString(), uuid, result));
 
-                } catch (LibvirtException e) {
-                    s_logger.debug(String.format("Failed to lookup libvirt storage pool [%s].", uuid), e);
+                        if (result != null) {
+                            s_logger.warn(String.format("Write heartbeat for pool [%s] failed: %s; try: %s of %s.", uuid, result, i, _heartBeatUpdateMaxTries));
+                            try {
+                                Thread.sleep(_heartBeatUpdateRetrySleep);
+                            } catch (InterruptedException e) {
+                                s_logger.debug("[IGNORED] Interrupted between heartbeat retries.", e);
+                            }
+                        } else {
+                            break;
+                        }
 
-                    if (e.toString().contains("pool not found")) {
-                        s_logger.debug(String.format("Removing pool [%s] from HA monitor since it was deleted.", uuid));
-                        removedPools.add(uuid);
-                        continue;
                     }
 
+                    if (result != null && rebootHostAndAlertManagementOnHeartbeatTimeout) {
+                        s_logger.warn(String.format("Write heartbeat for pool [%s] failed: %s; stopping cloudstack-agent.", uuid, result));
+                        Script cmd = createHeartBeatCommand(primaryStoragePool, null, false);
+                        result = cmd.execute();
+                    }
                 }
 
-                String result = null;
-                for (int i = 1; i <= _heartBeatUpdateMaxTries; i++) {
-                    Script cmd = createHeartBeatCommand(primaryStoragePool, hostPrivateIp, true);
-                    result = cmd.execute();
+                if (!removedPools.isEmpty()) {
+                    for (String uuid : removedPools) {
+                        removeStoragePool(uuid);
+                    }
+                }
+            }
+        }
 
-                    s_logger.debug(String.format("The command (%s), to the pool [%s], has the result [%s].", cmd.toString(), uuid, result));
+        if (rbdstoragePool != null && !rbdstoragePool.isEmpty()) {
```

Comment:
This code and the one above look quite similar, we could extract it to a method

Reply:
@GutoVeronezi  OK, I made method 'run

Reply:
I agree with @GutoVeronezi and you could do more than just the `runHeartbeatToPool()`. The rest of the block can be parameterised with the synchronisation object, `nfsstoragePool` and `rbdstoragePool` respectively. `obj.getClass()` to get the type and instantiate the and get the object in the loop, or add the type ars parameter as well.

no blocker, just a suggestion!